--- a/ipq806x/nss_gmac_dev.c
+++ b/ipq806x/nss_gmac_dev.c
@@ -1585,7 +1585,7 @@ int32_t nss_gmac_attach(struct nss_gmac_
 	}
 
 	/* ioremap addresses */
-	gmacdev->mac_base = ioremap_nocache(reg_base,
+	gmacdev->mac_base = ioremap(reg_base,
 						      NSS_GMAC_REG_BLOCK_LEN);
 	if (!gmacdev->mac_base) {
 		netdev_dbg(netdev, "ioremap fail.\n");
--- a/ipq806x/nss_gmac_ctrl.c
+++ b/ipq806x/nss_gmac_ctrl.c
@@ -1096,7 +1096,7 @@ static int32_t nss_gmac_do_common_init(s
 	ctx.msm_clk_ctl_enabled = true;
 #endif
 
-	ctx.nss_base = (uint8_t *)ioremap_nocache(res_nss_base.start,
+	ctx.nss_base = (uint8_t *)ioremap(res_nss_base.start,
 						  resource_size(&res_nss_base));
 	if (!ctx.nss_base) {
 		pr_info("Error mapping NSS GMAC registers\n");
@@ -1105,7 +1105,7 @@ static int32_t nss_gmac_do_common_init(s
 	}
 	pr_debug("%s: NSS base ioremap OK.\n", __func__);
 
-	ctx.qsgmii_base = (uint32_t *)ioremap_nocache(res_qsgmii_base.start,
+	ctx.qsgmii_base = (uint32_t *)ioremap(res_qsgmii_base.start,
 					      resource_size(&res_qsgmii_base));
 	if (!ctx.qsgmii_base) {
 		pr_info("Error mapping QSGMII registers\n");
@@ -1115,7 +1115,7 @@ static int32_t nss_gmac_do_common_init(s
 	pr_debug("%s: QSGMII base ioremap OK, vaddr = 0x%p\n",
 						__func__, ctx.qsgmii_base);
 
-	ctx.clk_ctl_base = (uint32_t *)ioremap_nocache(res_clk_ctl_base.start,
+	ctx.clk_ctl_base = (uint32_t *)ioremap(res_clk_ctl_base.start,
 				       resource_size(&res_clk_ctl_base));
 	if (!ctx.clk_ctl_base) {
 		pr_info("Error mapping Clk control registers\n");
