--- a/netlink/Makefile
+++ b/netlink/Makefile
@@ -1,3 +1,4 @@
+GRE_ENABLED := $(strip $(if $(filter $(gre), y), 1 , 0))
 CAPWAP_ENABLED := $(strip $(if $(filter $(capwapmgr), y), 1 , 0))
 IPSEC_ENABLED := $(strip $(if $(filter $(ipsecmgr), y), 1 , 0))
 DTLS_ENABLED := $(strip $(if $(filter $(dtlsmgr), y), 1 , 0))
@@ -10,44 +11,39 @@ ccflags-y += -DNSS_CLIENT_BUILD_ID="$(BU
 
 ccflags-y += -DCONFIG_NSS_NLIPV4=1
 ccflags-y += -DCONFIG_NSS_NLIPV6=1
-ccflags-y += -DCONFIG_NSS_NLOAM=1
-ccflags-y += -DCONFIG_NSS_NLGRE_REDIR_FAMILY=1
+ccflags-y += -DCONFIG_NSS_NLOAM=0
+ccflags-y += -DCONFIG_NSS_NLGRE_REDIR_FAMILY=${GRE_ENABLED}
 ccflags-y += -DCONFIG_NSS_NLETHRX=1
 ccflags-y += -DCONFIG_NSS_NLDYNAMIC_INTERFACE=1
 ccflags-y += -DCONFIG_NSS_NLN2H=1
-ccflags-y += -DCONFIG_NSS_NLIPV4_REASM=1
-ccflags-y += -DCONFIG_NSS_NLIPV6_REASM=1
-ccflags-y += -DCONFIG_NSS_NLWIFILI=1
-ccflags-y += -DCONFIG_NSS_NLLSO_RX=1
-ccflags-y += -DCONFIG_NSS_NLMAP_T=1
-ccflags-y += -DCONFIG_NSS_NLPPPOE=1
-ccflags-y += -DCONFIG_NSS_NLL2TPV2=1
-ccflags-y += -DCONFIG_NSS_NLPPTP=1
+ccflags-y += -DCONFIG_NSS_NLIPV4_REASM=0
+ccflags-y += -DCONFIG_NSS_NLIPV6_REASM=0
+ccflags-y += -DCONFIG_NSS_NLWIFILI=0
+ccflags-y += -DCONFIG_NSS_NLLSO_RX=0
+ccflags-y += -DCONFIG_NSS_NLMAP_T=0
+ccflags-y += -DCONFIG_NSS_NLPPPOE=0
+ccflags-y += -DCONFIG_NSS_NLL2TPV2=0
+ccflags-y += -DCONFIG_NSS_NLPPTP=0
 ccflags-y += -DCONFIG_NSS_NLCAPWAP=${CAPWAP_ENABLED}
 ccflags-y += -DCONFIG_NSS_NLIPSEC=${IPSEC_ENABLED}
 ccflags-y += -DCONFIG_NSS_NLDTLS=${DTLS_ENABLED}
 
 qca-nss-netlink-objs := nss_nl.o
-qca-nss-netlink-objs += nss_nlgre_redir_family.o
-qca-nss-netlink-objs += nss_nlgre_redir_cmd.o
-qca-nss-netlink-objs += nss_nlgre_redir_cmn.o
-qca-nss-netlink-objs += nss_nlgre_redir.o
-qca-nss-netlink-objs += nss_nlgre_redir_lag.o
 qca-nss-netlink-objs += nss_nlipv4.o
 qca-nss-netlink-objs += nss_nlipv6.o
-qca-nss-netlink-objs += nss_nloam.o
+# qca-nss-netlink-objs += nss_nloam.o
 qca-nss-netlink-objs += nss_nlethrx.o
 qca-nss-netlink-objs += nss_nldynamic_interface.o
 qca-nss-netlink-objs += nss_nln2h.o
-qca-nss-netlink-objs += nss_nlipv4_reasm.o
-qca-nss-netlink-objs += nss_nlipv6_reasm.o
-qca-nss-netlink-objs += nss_nlwifili.o
+# qca-nss-netlink-objs += nss_nlipv4_reasm.o
+# qca-nss-netlink-objs += nss_nlipv6_reasm.o
+# qca-nss-netlink-objs += nss_nlwifili.o
 qca-nss-netlink-objs += nss_nllso_rx.o
-qca-nss-netlink-objs += nss_nlmap_t.o
-qca-nss-netlink-objs += nss_nlpppoe.o
-qca-nss-netlink-objs += nss_nll2tpv2.o
-qca-nss-netlink-objs += nss_nlpptp.o
-
+# qca-nss-netlink-objs += nss_nlmap_t.o
+# qca-nss-netlink-objs += nss_nlpppoe.o
+# qca-nss-netlink-objs += nss_nll2tpv2.o
+# qca-nss-netlink-objs += nss_nlpptp.o
+#
 ifneq (,$(filter $(capwapmgr), y))
 qca-nss-netlink-objs += nss_nlcapwap.o
 endif
@@ -60,6 +56,14 @@ ifneq (,$(filter $(ipsecmgr), y))
 qca-nss-netlink-objs += nss_nlipsec.o
 endif
 
+ifneq (,$(filter $(gre), y))
+qca-nss-netlink-objs += nss_nlgre_redir_family.o
+qca-nss-netlink-objs += nss_nlgre_redir_cmd.o
+qca-nss-netlink-objs += nss_nlgre_redir_cmn.o
+qca-nss-netlink-objs += nss_nlgre_redir.o
+qca-nss-netlink-objs += nss_nlgre_redir_lag.o
+endif
+
 ifeq ($(SoC),$(filter $(SoC),ipq807x ipq807x_64))
 ccflags-y += -DCONFIG_NSS_NLC2C_TX=1
 ccflags-y += -DCONFIG_NSS_NLC2C_RX=1
--- a/netlink/nss_nl.c
+++ b/netlink/nss_nl.c
@@ -106,24 +106,6 @@ static struct nss_nl_family family_handl
 	},
 	{
 		/*
-		 * NSS_NLIPSEC
-		 */
-		.name = NSS_NLIPSEC_FAMILY,		/* ipsec */
-		.entry = NSS_NLIPSEC_INIT,		/* init */
-		.exit = NSS_NLIPSEC_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLIPSEC		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLOAM
-		 */
-		.name = NSS_NLOAM_FAMILY,		/* oam */
-		.entry = NSS_NLOAM_INIT,		/* init */
-		.exit = NSS_NLOAM_EXIT,			/* exit */
-		.valid = CONFIG_NSS_NLOAM		/* 1 or 0 */
-	},
-	{
-		/*
 		 * NSS_NLIPV6
 		 */
 		.name = NSS_NLIPV6_FAMILY,		/* ipv6 */
@@ -133,33 +115,6 @@ static struct nss_nl_family family_handl
 	},
 	{
 		/*
-		 * NSS_NLGRE_REDIR
-		 */
-		.name = NSS_NLGRE_REDIR_FAMILY,		/* gre_redir */
-		.entry = NSS_NLGRE_REDIR_FAMILY_INIT,	/* init */
-		.exit = NSS_NLGRE_REDIR_FAMILY_EXIT,	/* exit */
-		.valid = CONFIG_NSS_NLGRE_REDIR_FAMILY	/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLCAPWAP
-		 */
-		.name = NSS_NLCAPWAP_FAMILY,		/* capwap */
-		.entry = NSS_NLCAPWAP_INIT,		/* init */
-		.exit = NSS_NLCAPWAP_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLCAPWAP		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLDTLS
-		 */
-		.name = NSS_NLDTLS_FAMILY,		/* dtls */
-		.entry = NSS_NLDTLS_INIT,		/* init */
-		.exit = NSS_NLDTLS_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLDTLS		/* 1 or 0 */
-	},
-	{
-		/*
 		 * NSS_NLETHRX
 		 */
 		.name = NSS_NLETHRX_FAMILY,		/* ethrx */
@@ -169,15 +124,6 @@ static struct nss_nl_family family_handl
 	},
 	{
 		/*
-		 * NSS_NLEDMA
-		 */
-		.name = NSS_NLEDMA_FAMILY,		/* edma */
-		.entry = NSS_NLEDMA_INIT,		/* init */
-		.exit = NSS_NLEDMA_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLEDMA		/* 1 or 0 */
-	},
-	{
-		/*
 		 * NSS_NLDYNAMIC_INTERFACE
 		 */
 		.name = NSS_NLDYNAMIC_INTERFACE_FAMILY,	/* dynamic interface */
@@ -196,51 +142,6 @@ static struct nss_nl_family family_handl
 	},
 	{
 		/*
-		 * NSS_NLC2C_TX
-		 */
-		.name = NSS_NLC2C_TX_FAMILY,		/* c2c_tx */
-		.entry = NSS_NLC2C_TX_INIT,		/* init */
-		.exit = NSS_NLC2C_TX_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLC2C_TX		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLC2C_RX
-		 */
-		.name = NSS_NLC2C_RX_FAMILY,		/* c2c_rx */
-		.entry = NSS_NLC2C_RX_INIT,		/* init */
-		.exit = NSS_NLC2C_RX_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLC2C_RX		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLIPV4_REASM
-		 */
-		.name = NSS_NLIPV4_REASM_FAMILY,	/* ipv4_reasm */
-		.entry = NSS_NLIPV4_REASM_INIT,		/* init */
-		.exit = NSS_NLIPV4_REASM_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLIPV4_REASM	/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLIPV6_REASM
-		 */
-		.name = NSS_NLIPV6_REASM_FAMILY,	/* ipv6_reasm */
-		.entry = NSS_NLIPV6_REASM_INIT,		/* init */
-		.exit = NSS_NLIPV6_REASM_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLIPV6_REASM	/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLWIFILI
-		 */
-		.name = NSS_NLWIFILI_FAMILY,		/* wifili */
-		.entry = NSS_NLWIFILI_INIT,		/* init */
-		.exit = NSS_NLWIFILI_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLWIFILI		/* 1 or 0 */
-	},
-	{
-		/*
 		 * NSS_NLLSO_RX
 		 */
 		.name = NSS_NLLSO_RX_FAMILY,		/* lso_rx */
@@ -248,42 +149,6 @@ static struct nss_nl_family family_handl
 		.exit = NSS_NLLSO_RX_EXIT,		/* exit */
 		.valid = CONFIG_NSS_NLLSO_RX		/* 1 or 0 */
 	},
-	{
-		/*
-		 * NSS_NLMAP_T
-		 */
-		.name = NSS_NLMAP_T_FAMILY,		/* map_t */
-		.entry = NSS_NLMAP_T_INIT,		/* init */
-		.exit = NSS_NLMAP_T_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLMAP_T		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLPPPOE
-		 */
-		.name = NSS_NLPPPOE_FAMILY,		/* pppoe */
-		.entry = NSS_NLPPPOE_INIT,		/* init */
-		.exit = NSS_NLPPPOE_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLPPPOE		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLL2TPV2
-		 */
-		.name = NSS_NLL2TPV2_FAMILY,		/* l2tpv2 */
-		.entry = NSS_NLL2TPV2_INIT,		/* init */
-		.exit = NSS_NLL2TPV2_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLL2TPV2		/* 1 or 0 */
-	},
-	{
-		/*
-		 * NSS_NLPPTP
-		 */
-		.name = NSS_NLPPTP_FAMILY,		/* pptp */
-		.entry = NSS_NLPPTP_INIT,		/* init */
-		.exit = NSS_NLPPTP_EXIT,		/* exit */
-		.valid = CONFIG_NSS_NLPPTP		/* 1 or 0 */
-	},
 };
 
 #define NSS_NL_FAMILY_HANDLER_SZ ARRAY_SIZE(family_handlers)
--- a/netlink/nss_nlipv4.c
+++ b/netlink/nss_nlipv4.c
@@ -336,20 +336,6 @@ static int nss_nlipv4_verify_conn_rule(s
 							tuple->return_ident, tuple->flow_ident);
 		break;
 
-	case NSS_NL_IFTYPE_TUNNEL_GRE:
-		/*
-		 * Currently this implementation is only for gre_redir
-		 */
-		conn->flow_interface_num = nss_nlgre_redir_cmd_get_ifnum(flow_dev, tuple->protocol);
-		if (conn->flow_interface_num < 0 ) {
-			nss_nl_error("%px: Failed to get flow interface number (dev:%s, type:%d)\n",
-								flow_dev, flow_dev->name, flow_iftype);
-			return -EINVAL;
-		}
-
-		conn->flow_mtu = nss_nlgre_redir_cmd_get_mtu(flow_dev, NSS_GRE_REDIR_IP_HDR_TYPE_IPV4, conn->flow_interface_num);
-		break;
-
 	case NSS_NL_IFTYPE_VLAN:
 		conn->flow_interface_num = nss_cmn_get_interface_number_by_dev(vlan_dev_real_dev(flow_dev));
 		if (conn->flow_interface_num < 0 ) {
@@ -396,17 +382,6 @@ static int nss_nlipv4_verify_conn_rule(s
 							tuple->return_ident, tuple->flow_ident);
 		break;
 
-	case NSS_NL_IFTYPE_TUNNEL_GRE:
-		conn->return_interface_num = nss_nlgre_redir_cmd_get_ifnum(return_dev, tuple->protocol);
-		if (conn->return_interface_num < 0 ) {
-			nss_nl_error("%px: Failed to get return interface number (dev:%s, type:%d)\n",
-							return_dev, return_dev->name, return_iftype);
-			return -EINVAL;
-		}
-
-		conn->return_mtu = nss_nlgre_redir_cmd_get_mtu(return_dev, NSS_GRE_REDIR_IP_HDR_TYPE_IPV4, conn->return_interface_num);
-		break;
-
 	case NSS_NL_IFTYPE_VLAN:
 		conn->return_interface_num = nss_cmn_get_interface_number_by_dev(vlan_dev_real_dev(return_dev));
 		if (conn->return_interface_num < 0 ) {
--- a/netlink/nss_nlipv6.c
+++ b/netlink/nss_nlipv6.c
@@ -353,17 +353,6 @@ static int nss_nlipv6_verify_conn_rule(s
 							tuple->return_ident, tuple->flow_ident);
 		break;
 
-	case NSS_NL_IFTYPE_TUNNEL_GRE:
-		conn->flow_interface_num = nss_nlgre_redir_cmd_get_ifnum(flow_dev, tuple->protocol);
-		if (conn->flow_interface_num < 0 ) {
-			nss_nl_error("%px: Failed to get flow interface number (dev:%s, type:%d)\n",
-			flow_dev, flow_dev->name, flow_iftype);
-			return -EINVAL;
-		}
-
-		conn->flow_mtu = nss_nlgre_redir_cmd_get_mtu(flow_dev, NSS_GRE_REDIR_IP_HDR_TYPE_IPV6, conn->flow_interface_num);
-		break;
-
 	case NSS_NL_IFTYPE_VLAN:
 		conn->flow_interface_num = nss_cmn_get_interface_number_by_dev(vlan_dev_real_dev(flow_dev));
 		if (conn->flow_interface_num < 0 ) {
@@ -411,17 +400,6 @@ static int nss_nlipv6_verify_conn_rule(s
 							tuple->return_ident, tuple->flow_ident);
 		break;
 
-	case NSS_NL_IFTYPE_TUNNEL_GRE:
-		conn->return_interface_num = nss_nlgre_redir_cmd_get_ifnum(return_dev, tuple->protocol);
-		if (conn->return_interface_num < 0 ) {
-			nss_nl_error("%px: Failed to get return interface number (dev:%s, type:%d)\n",
-			return_dev, return_dev->name, return_iftype);
-			return -EINVAL;
-		}
-
-		conn->return_mtu = nss_nlgre_redir_cmd_get_mtu(return_dev, NSS_GRE_REDIR_IP_HDR_TYPE_IPV6, conn->return_interface_num);
-		break;
-
 	case NSS_NL_IFTYPE_VLAN:
 		conn->return_interface_num = nss_cmn_get_interface_number_by_dev(vlan_dev_real_dev(return_dev));
 		if (conn->return_interface_num < 0 ) {
