--- a/exports/nss_dynamic_interface.h
+++ b/exports/nss_dynamic_interface.h
@@ -22,6 +22,8 @@
 #ifndef __NSS_DYNAMIC_INTERFACE_H
 #define __NSS_DYNAMIC_INTERFACE_H
 
+#include "nss_fw_version.h"
+
 /**
  * @addtogroup nss_dynamic_interface_subsystem
  * @{
--- a/exports/nss_ipv4.h
+++ b/exports/nss_ipv4.h
@@ -27,6 +27,8 @@
  * @{
  */
 
+#include "nss_fw_version.h"
+
 /**
  * nss_ipv4_message_types
  *	IPv4 bridge and routing rule message types.
