--- a/nss_stats.c
+++ b/nss_stats.c
@@ -398,11 +398,6 @@ void nss_stats_create_dentry(char *name,
  */
 
 /*
- * gmac_stats_ops
- */
-NSS_STATS_DECLARE_FILE_OPERATIONS(gmac);
-
-/*
  * wt_stats_ops
  */
 NSS_STATS_DECLARE_FILE_OPERATIONS(wt);
@@ -453,11 +448,6 @@ void nss_stats_init(void)
 	nss_drv_stats_dentry_create();
 
 	/*
-	 * gmac_stats
-	 */
-	nss_stats_create_dentry("gmac", &nss_gmac_stats_ops);
-
-	/*
 	 * Per-project stats
 	 */
 	nss_top_main.project_dentry = debugfs_create_dir("project",
