--- a/Makefile
+++ b/Makefile
@@ -5,11 +5,12 @@
 obj-m += qca-nss-drv.o
 
 #
+# Regular NSS path
+#
+#
 # List the files that belong to the driver in alphabetical order.
 #
 qca-nss-drv-objs := \
-			nss_bridge.o \
-			nss_bridge_log.o \
 			nss_cmn.o \
 			nss_core.o \
 			nss_coredump.o \
@@ -18,74 +19,70 @@ qca-nss-drv-objs := \
 			nss_dynamic_interface.o \
 			nss_dynamic_interface_log.o \
 			nss_dynamic_interface_stats.o \
-			nss_eth_rx.o \
-			nss_eth_rx_stats.o \
-			nss_eth_rx_strings.o \
-			nss_gmac_stats.o \
-			nss_if.o \
-			nss_if_log.o \
 			nss_init.o \
-			nss_ipv4.o \
-			nss_ipv4_stats.o \
-			nss_ipv4_strings.o \
-			nss_ipv4_log.o \
-			nss_ipv4_reasm.o \
-			nss_ipv4_reasm_stats.o \
-			nss_ipv4_reasm_strings.o \
-			nss_ipv6.o \
-			nss_ipv6_stats.o \
-			nss_ipv6_strings.o \
-			nss_ipv6_log.o \
-			nss_ipv6_reasm.o \
-			nss_ipv6_reasm_stats.o \
-			nss_ipv6_reasm_strings.o \
 			nss_log.o \
-			nss_lso_rx.o \
-			nss_lso_rx_stats.o \
-			nss_lso_rx_strings.o \
 			nss_meminfo.o \
 			nss_n2h.o \
 			nss_n2h_stats.o \
 			nss_n2h_strings.o \
-			nss_phys_if.o \
 			nss_pm.o \
 			nss_profiler.o \
 			nss_project.o \
-			nss_pppoe.o \
-			nss_pppoe_log.o \
-			nss_pppoe_stats.o \
-			nss_pppoe_strings.o \
 			nss_rps.o \
 			nss_stats.o \
 			nss_strings.o \
-			nss_tx_msg_sync.o \
 			nss_unaligned.o \
 			nss_unaligned_log.o \
-			nss_unaligned_stats.o \
+			nss_unaligned_stats.o
+
+# Base NSS HAL support
+qca-nss-drv-objs += nss_hal/nss_hal.o
+
+ifneq "$(NSS_DRV_POINT_OFFLOAD)" "y"
+qca-nss-drv-objs += \
+			nss_gmac_stats.o \
+			nss_if.o \
+			nss_if_log.o \
+			nss_phys_if.o \
+			nss_tx_msg_sync.o \
 			nss_virt_if.o \
-			nss_virt_if_stats.o \
-			nss_vlan.o \
-			nss_vlan_log.o \
-			nss_wifi.o \
-			nss_wifi_log.o \
-			nss_wifi_stats.o \
-			nss_wifi_vdev.o \
-			nss_wifili.o \
-			nss_wifili_log.o \
-			nss_wifili_stats.o \
-			nss_wifili_strings.o \
-			nss_wifi_mac_db.o \
-			nss_wifi_ext_vdev.o \
-			nss_wifi_ext_vdev_stats.o \
-			nss_wifi_ext_vdev_log.o \
-			nss_wifi_mesh.o \
-			nss_wifi_mesh_log.o \
-			nss_wifi_mesh_stats.o \
-			nss_wifi_mesh_strings.o
+			nss_virt_if_stats.o
 
-# Base NSS data plane/HAL support
+# Base NSS data plane support
 qca-nss-drv-objs += nss_data_plane/nss_data_plane_common.o
-qca-nss-drv-objs += nss_hal/nss_hal.o
+endif
+
+ifneq "$(NSS_DRV_BRIDGE_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_BRIDGE_ENABLE
+qca-nss-drv-objs += \
+		    nss_bridge.o \
+		    nss_bridge_log.o
+endif
+
+ifneq "$(NSS_DRV_WIFI_EXT_VDEV_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFI_EXT_VDEV_ENABLE
+qca-nss-drv-objs += \
+		    nss_wifi_ext_vdev.o \
+		    nss_wifi_ext_vdev_stats.o \
+		    nss_wifi_ext_vdev_log.o
+endif
+
+ifneq "$(NSS_DRV_WIFI_MESH_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFI_MESH_ENABLE
+qca-nss-drv-objs += \
+		    nss_wifi_mesh.o \
+		    nss_wifi_mesh_log.o \
+		    nss_wifi_mesh_stats.o \
+		    nss_wifi_mesh_strings.o
+endif
+
+ifneq "$(NSS_DRV_IPV4_REASM_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_IPV4_REASM_ENABLE
+qca-nss-drv-objs += \
+		    nss_ipv4_reasm.o \
+		    nss_ipv4_reasm_stats.o \
+		    nss_ipv4_reasm_strings.o
+endif
 
 ifneq "$(NSS_DRV_L2TP_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_L2TP_ENABLE
@@ -111,6 +108,22 @@ qca-nss-drv-objs += \
 			 nss_pvxlan_stats.o
 endif
 
+ifneq "$(NSS_DRV_IPV6_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_IPV6_ENABLE
+qca-nss-drv-objs += \
+			nss_ipv6.o \
+			nss_ipv6_stats.o \
+			nss_ipv6_strings.o \
+			nss_ipv6_log.o
+ifneq "$(NSS_DRV_IPV6_REASM_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_IPV6_REASM_ENABLE
+qca-nss-drv-objs += \
+		    nss_ipv6_reasm.o \
+		    nss_ipv6_reasm_stats.o \
+		    nss_ipv6_reasm_strings.o
+endif
+endif
+
 ifneq "$(NSS_DRV_TSTAMP_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_TSTAMP_ENABLE
 qca-nss-drv-objs += \
@@ -213,6 +226,14 @@ qca-nss-drv-objs += \
 			 nss_trustsec_tx_stats.o
 endif
 
+ifneq "$(NSS_DRV_TRUSTSEC_RX_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_TRUSTSEC_RX_ENABLE
+qca-nss-drv-objs += \
+			 nss_trustsec_rx.o \
+			 nss_trustsec_rx_log.o \
+			 nss_trustsec_rx_stats.o
+endif
+
 ifneq "$(NSS_DRV_TUNIPIP6_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_TUNIPIP6_ENABLE
 qca-nss-drv-objs += \
@@ -303,7 +324,70 @@ qca-nss-drv-objs += \
 			nss_udp_st_strings.o
 endif
 
+ifneq "$(NSS_DRV_IPV4_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_IPV4_ENABLE
+qca-nss-drv-objs += \
+			nss_ipv4.o \
+			nss_ipv4_stats.o \
+			nss_ipv4_strings.o \
+			nss_ipv4_log.o
+endif
+
+ifneq "$(NSS_DRV_ETH_RX_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_ETH_RX_ENABLE
+qca-nss-drv-objs += \
+			nss_eth_rx.o \
+			nss_eth_rx_stats.o \
+			nss_eth_rx_strings.o
+endif
+
+ifneq "$(NSS_DRV_PPPOE_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_PPPOE_ENABLE
+qca-nss-drv-objs += \
+			nss_pppoe.o \
+			nss_pppoe_log.o \
+			nss_pppoe_stats.o \
+			nss_pppoe_strings.o
+endif
+
+ifneq "$(NSS_DRV_WIFIOFFLOAD_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFIOFFLOAD_ENABLE
+ifneq "$(NSS_DRV_WIFI_LEGACY_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFI_LEGACY_ENABLE
+qca-nss-drv-objs += \
+			nss_wifi.o \
+			nss_wifi_log.o \
+			nss_wifi_stats.o
+ifeq ($(SoC),$(filter $(SoC),ipq806x))
+ccflags-y += -DNSS_HAL_IPQ806x_SUPPORT
+endif
+endif
+qca-nss-drv-objs += \
+			nss_wifi_vdev.o \
+			nss_wifili.o \
+			nss_wifili_log.o \
+			nss_wifili_stats.o \
+			nss_wifili_strings.o \
+			nss_wifi_mac_db.o
+endif
+
+ifneq "$(NSS_DRV_VLAN_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_VLAN_ENABLE
+qca-nss-drv-objs += \
+			nss_vlan.o \
+			nss_vlan_log.o
+endif
+
+ifneq "$(NSS_DRV_LSO_RX_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_LSO_RX_ENABLE
+qca-nss-drv-objs += \
+			nss_lso_rx.o \
+			nss_lso_rx_stats.o \
+			nss_lso_rx_strings.o
+endif
+
 ifeq ($(SoC),$(filter $(SoC),ipq806x))
+ccflags-y += -DNSS_DRV_IPV4_ENABLE -DNSS_DRV_IPV6_ENABLE
 qca-nss-drv-objs += nss_data_plane/nss_data_plane_gmac.o \
 		    nss_hal/ipq806x/nss_hal_pvt.o
 
@@ -409,6 +493,51 @@ qca-nss-drv-objs += \
 endif
 endif
 
+ifeq ($(SoC),$(filter $(SoC),ipq95xx))
+qca-nss-drv-objs += nss_hal/ipq95xx/nss_hal_pvt.o
+qca-nss-drv-objs += nss_data_plane/nss_data_plane_lite.o
+qca-nss-drv-objs += nss_edma_lite.o \
+		    nss_edma_lite_stats.o \
+		    nss_edma_lite_strings.o
+
+ccflags-y += -DNSS_DATA_PLANE_LITE_SUPPORT
+ccflags-y += -DNSS_DRV_EDMA_LITE_ENABLE
+ccflags-y += -DNSS_DRV_POINT_OFFLOAD
+
+ifneq "$(NSS_DRV_CRYPTO_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_CRYPTO_ENABLE
+ccflags-y += -DNSS_DRV_DMA_ENABLE
+
+qca-nss-drv-objs += \
+		    nss_crypto.o \
+		    nss_crypto_log.o \
+		    nss_crypto_cmn.o \
+		    nss_crypto_cmn_log.o \
+		    nss_crypto_cmn_stats.o \
+		    nss_crypto_cmn_strings.o \
+		    nss_dma.o \
+		    nss_dma_log.o \
+		    nss_dma_stats.o \
+		    nss_dma_strings.o
+endif
+
+ifneq "$(NSS_DRV_DTLS_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_DTLS_ENABLE
+qca-nss-drv-objs += \
+		    nss_dtls.o \
+		    nss_dtls_log.o \
+		    nss_dtls_stats.o \
+		    nss_dtls_cmn.o \
+		    nss_dtls_cmn_log.o \
+		    nss_dtls_cmn_stats.o \
+		    nss_dtls_cmn_strings.o
+endif
+
+ccflags-y += -I$(obj)/nss_hal/ipq95xx -DNSS_HAL_IPQ95XX_SUPPORT -DNSS_MULTI_H2N_DATA_RING_SUPPORT
+else
+ccflags-y += -DNSS_DATA_PLANE_GENERIC_SUPPORT
+endif
+
 ifeq ($(SoC),$(filter $(SoC),ipq807x ipq807x_64))
 qca-nss-drv-objs += nss_hal/ipq807x/nss_hal_pvt.o \
 		    nss_data_plane/hal/nss_ipq807x.o
@@ -467,6 +596,7 @@ ccflags-y += -I$(obj)/nss_hal/ipq50xx -D
 endif
 
 ccflags-y += -I$(obj)/nss_hal/include -I$(obj)/nss_data_plane/include -I$(obj)/exports -DNSS_DEBUG_LEVEL=0 -DNSS_PKT_STATS_ENABLED=1
+
 ccflags-y += -I$(obj)/nss_data_plane/hal/include
 ccflags-y += -DNSS_PM_DEBUG_LEVEL=0 -DNSS_SKB_REUSE_SUPPORT=1
 ccflags-y += -Wall -Werror
@@ -481,7 +611,14 @@ NSS_CCFLAGS = -DNSS_DT_SUPPORT=1 -DNSS_F
 ccflags-y += -I$(obj)
 endif
 
+# Fabric scaling is supported in 3.14 and 4.4 only
+ifneq ($(findstring 3.14, $(KERNELVERSION)),)
+NSS_CCFLAGS += -DNSS_FABRIC_SCALING_SUPPORT=1
+else ifneq ($(findstring 4.4, $(KERNELVERSION)),)
+NSS_CCFLAGS += -DNSS_FABRIC_SCALING_SUPPORT=1
+else
 NSS_CCFLAGS += -DNSS_FABRIC_SCALING_SUPPORT=0
+endif
 
 # Disable Frequency scaling
 ifeq "$(NSS_FREQ_SCALE_DISABLE)" "y"
--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -65,9 +65,9 @@ int nss_hal_firmware_load(struct nss_ctx
 	int rc;
 
 	if (nss_ctx->id == 0) {
-		rc = request_firmware(&nss_fw, NSS_AP0_IMAGE, &(nss_dev->dev));
+		rc = firmware_request_nowarn(&nss_fw, NSS_AP0_IMAGE, &(nss_dev->dev));
 	} else if (nss_ctx->id == 1) {
-		rc = request_firmware(&nss_fw, NSS_AP1_IMAGE, &(nss_dev->dev));
+		rc = firmware_request_nowarn(&nss_fw, NSS_AP1_IMAGE, &(nss_dev->dev));
 	} else {
 		nss_warning("%px: Invalid nss dev: %d\n", nss_ctx, nss_ctx->id);
 		return -EINVAL;
--- a/nss_init.c
+++ b/nss_init.c
@@ -935,6 +935,13 @@ nss_info("Init NSS driver");
 #endif
 
 	/*
+	 * Init Wi-Fi mesh
+	 */
+#ifdef NSS_DRV_WIFI_MESH_ENABLE
+	nss_wifi_mesh_init();
+#endif
+
+	/*
 	 * Register platform_driver
 	 */
 	return platform_driver_register(&nss_driver);
