--- a/Makefile
+++ b/Makefile
@@ -46,13 +46,7 @@ qca-nss-drv-objs := \
 			nss_lso_rx.o \
 			nss_lso_rx_stats.o \
 			nss_lso_rx_strings.o \
-			nss_match.o \
-			nss_match_log.o \
-			nss_match_stats.o \
 			nss_meminfo.o \
-			nss_mirror.o \
-			nss_mirror_log.o \
-			nss_mirror_stats.o \
 			nss_n2h.o \
 			nss_n2h_stats.o \
 			nss_n2h_strings.o \
@@ -60,28 +54,44 @@ qca-nss-drv-objs := \
 			nss_pm.o \
 			nss_profiler.o \
 			nss_project.o \
-			nss_ppe.o \
-			nss_ppe_log.o \
-			nss_ppe_stats.o \
-			nss_ppe_vp.o \
-			nss_ppe_vp_log.o \
-			nss_ppe_vp_stats.o \
-			nss_pppoe.o \
-			nss_pppoe_log.o \
-			nss_pppoe_stats.o \
-			nss_pppoe_strings.o \
 			nss_rps.o \
 			nss_stats.o \
 			nss_strings.o \
 			nss_tx_msg_sync.o \
 			nss_unaligned.o \
 			nss_unaligned_log.o \
-			nss_unaligned_stats.o \
+			nss_unaligned_stats.o
+
+# Base NSS data plane/HAL support
+qca-nss-drv-objs += nss_data_plane/nss_data_plane_common.o
+qca-nss-drv-objs += nss_hal/nss_hal.o
+
+ifneq "$(NSS_DRV_PPPOE_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_PPPOE_ENABLE
+qca-nss-drv-objs += \
+			nss_pppoe.o \
+			nss_pppoe_log.o \
+			nss_pppoe_stats.o \
+			nss_pppoe_strings.o
+endif
+
+ifneq "$(NSS_DRV_VIRT_IF_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_VIRT_IF_ENABLE
+qca-nss-drv-objs += \
 			nss_virt_if.o \
-			nss_virt_if_stats.o \
+			nss_virt_if_stats.o
+endif
+
+ifneq "$(NSS_DRV_VLAN_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_VLAN_ENABLE
+qca-nss-drv-objs += \
 			nss_vlan.o \
-			nss_vlan_log.o \
-			nss_wifi.o \
+			nss_vlan_log.o
+endif
+
+ifneq "$(NSS_DRV_WIFI_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFI_ENABLE
+qca-nss-drv-objs += nss_wifi.o \
 			nss_wifi_log.o \
 			nss_wifi_stats.o \
 			nss_wifi_vdev.o \
@@ -92,25 +102,22 @@ qca-nss-drv-objs := \
 			nss_wifili_stats.o \
 			nss_wifili_strings.o \
 			nss_wifi_mac_db.o
-
-# Base NSS data plane/HAL support
-qca-nss-drv-objs += nss_data_plane/nss_data_plane_common.o
-qca-nss-drv-objs += nss_hal/nss_hal.o
+endif
 
 ifneq "$(NSS_DRV_L2TP_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_L2TP_ENABLE
 qca-nss-drv-objs += \
-		   nss_l2tpv2.o \
-		   nss_l2tpv2_log.o \
-		   nss_l2tpv2_stats.o \
-		   nss_l2tpv2_strings.o
+			nss_l2tpv2.o \
+			nss_l2tpv2_log.o \
+			nss_l2tpv2_stats.o \
+			nss_l2tpv2_strings.o
 endif
 
 ifneq "$(NSS_DRV_LAG_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_LAG_ENABLE
 qca-nss-drv-objs += \
-		   nss_lag.o \
-		   nss_lag_log.o
+			nss_lag.o \
+			nss_lag_log.o
 endif
 
 ifneq "$(NSS_DRV_PVXLAN_ENABLE)" "n"
@@ -121,6 +128,22 @@ qca-nss-drv-objs += \
 			 nss_pvxlan_stats.o
 endif
 
+ifneq "$(NSS_DRV_MATCH_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_MATCH_ENABLE
+qca-nss-drv-objs += \
+			nss_match.o \
+			nss_match_log.o \
+			nss_match_stats.o
+endif
+
+ifneq "$(NSS_DRV_MIRROR_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_MIRROR_ENABLE
+qca-nss-drv-objs += \
+			nss_mirror.o \
+			nss_mirror_log.o \
+			nss_mirror_stats.o
+endif
+
 ifneq "$(NSS_DRV_TSTAMP_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_TSTAMP_ENABLE
 qca-nss-drv-objs += \
@@ -292,6 +315,7 @@ endif
 ifeq ($(SoC),$(filter $(SoC),ipq806x ipq40xx))
 qca-nss-drv-objs += nss_data_plane/nss_data_plane_gmac.o \
 		    nss_hal/ipq806x/nss_hal_pvt.o
+
 ifneq "$(NSS_DRV_IPSEC_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_IPSEC_ENABLE
 qca-nss-drv-objs += \
@@ -317,7 +341,16 @@ ccflags-y += -I$(obj)/nss_hal/ipq806x -D
 endif
 
 ifeq ($(SoC),$(filter $(SoC),ipq60xx ipq60xx_64 ipq807x ipq807x_64))
-qca-nss-drv-objs += nss_data_plane/nss_data_plane.o
+qca-nss-drv-objs += nss_data_plane/nss_data_plane.o \
+			nss_ppe.o \
+			nss_ppe_log.o \
+			nss_ppe_stats.o \
+			nss_ppe_vp.o \
+			nss_ppe_vp_log.o \
+			nss_ppe_vp_stats.o
+
+ccflags-y += -DNSS_DRV_PPE_ENABLE
+ccflags-y += -DNSS_DRV_EDMA_ENABLE
 
 ifneq "$(NSS_DRV_IPSEC_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_IPSEC_ENABLE
--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -458,16 +458,20 @@ int nss_hal_probe(struct platform_device
 	}
 #endif
 
+#ifdef NSS_DRV_PPPOE_ENABLE
 	if (npd->pppoe_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->pppoe_handler_id = nss_dev->id;
 		nss_pppoe_register_handler();
 	}
+#endif
 
+#ifdef NSS_DRV_PPE_ENABLE
 	if (npd->ppe_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->ppe_handler_id = nss_dev->id;
 		nss_ppe_register_handler();
 		nss_ppe_vp_register_handler();
 	}
+#endif
 
 #ifdef NSS_DRV_L2TP_ENABLE
 	if (npd->l2tpv2_enabled == NSS_FEATURE_ENABLED) {
@@ -554,6 +558,7 @@ int nss_hal_probe(struct platform_device
 	}
 #endif
 
+#ifdef NSS_DRV_WIFI_ENABLE
 	if (npd->wifioffload_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->wifi_handler_id = nss_dev->id;
 		nss_top->dynamic_interface_table[NSS_DYNAMIC_INTERFACE_TYPE_VAP] = nss_dev->id;
@@ -569,6 +574,7 @@ int nss_hal_probe(struct platform_device
 		nss_top->wmdb_handler_id = nss_dev->id;
 		nss_wifi_mac_db_register_handler();
 	}
+#endif
 
 #ifdef NSS_DRV_OAM_ENABLE
 	if (npd->oam_enabled == NSS_FEATURE_ENABLED) {
@@ -583,11 +589,13 @@ int nss_hal_probe(struct platform_device
 		nss_bridge_init();
 	}
 
+#ifdef NSS_DRV_VLAN_ENABLE
 	if (npd->vlan_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->vlan_handler_id = nss_dev->id;
 		nss_top->dynamic_interface_table[NSS_DYNAMIC_INTERFACE_TYPE_VLAN] = nss_dev->id;
 		nss_vlan_register_handler();
 	}
+#endif
 
 #ifdef NSS_DRV_QVPN_ENABLE
 #if defined(NSS_HAL_IPQ807x_SUPPORT) || defined(NSS_HAL_IPQ60XX_SUPPORT)
@@ -641,11 +649,13 @@ int nss_hal_probe(struct platform_device
 	}
 #endif
 
+#ifdef NSS_DRV_MATCH_ENABLE
 	if (npd->match_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->match_handler_id = nss_dev->id;
 		nss_top->dynamic_interface_table[NSS_DYNAMIC_INTERFACE_TYPE_MATCH] = nss_dev->id;
 		nss_match_init();
 	}
+#endif
 
 #if (NSS_FW_VERSION_CODE > NSS_FW_VERSION(11,1))
 #ifdef NSS_DRV_TLS_ENABLE
--- a/nss_init.c
+++ b/nss_init.c
@@ -771,11 +771,12 @@ static int __init nss_init(void)
 	 */
 	nss_project_register_sysctl();
 
+#ifdef NSS_DRV_PPPOE_ENABLE
 	/*
 	 * Registering sysctl for pppoe specific config.
 	 */
 	nss_pppoe_register_sysctl();
-
+#endif
 	/*
 	 * Setup Runtime Sample values
 	 */
@@ -860,9 +861,9 @@ static int __init nss_init(void)
 	/*
 	 * INIT ppe on supported platform
 	 */
-	if (of_machine_is_compatible("qcom,ipq807x") || of_machine_is_compatible("qcom,ipq8074")|| of_machine_is_compatible("qcom,ipq6018")) {
-		nss_ppe_init();
-	}
+#ifdef NSS_DRV_PPE_ENABLE
+	nss_ppe_init();
+#endif
 
 #ifdef NSS_DRV_DMA_ENABLE
 	nss_dma_init();
@@ -902,10 +903,12 @@ static void __exit nss_cleanup(void)
 	nss_c2c_tx_unregister_sysctl();
 #endif
 
+#ifdef NSS_DRV_PPPOE_ENABLE
 	/*
 	 * Unregister pppoe specific sysctl
 	 */
 	nss_pppoe_unregister_sysctl();
+#endif
 
 	/*
 	 * Unregister ipv4/6 specific sysctl
@@ -925,9 +928,9 @@ static void __exit nss_cleanup(void)
 	/*
 	 * cleanup ppe on supported platform
 	 */
-	if (of_machine_is_compatible("qcom,ipq807x") || of_machine_is_compatible("qcom,ipq6018")) {
-		nss_ppe_free();
-	}
+#ifdef NSS_DRV_PPE_ENABLE
+	nss_ppe_free();
+#endif
 
 	platform_driver_unregister(&nss_driver);
 }
