--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -43,7 +43,6 @@
  */
 #define MIN_IMG_SIZE (64*1024)
 #define NSS_AP0_IMAGE "qca-nss0.bin"
-#define NSS_AP1_IMAGE "qca-nss1.bin"
 
 /*
  * File local/Static variables/functions
@@ -57,11 +56,9 @@ int nss_hal_firmware_load(struct nss_ctx
 	void __iomem *load_mem;
 	int rc;
 
-	if (nss_ctx->id == 0) {
+	if (nss_ctx->id == 0)
 		rc = request_firmware(&nss_fw, NSS_AP0_IMAGE, &(nss_dev->dev));
-	} else if (nss_ctx->id == 1) {
-		rc = request_firmware(&nss_fw, NSS_AP1_IMAGE, &(nss_dev->dev));
-	} else {
+	else {
 		nss_warning("%px: Invalid nss dev: %d\n", nss_ctx, nss_ctx->id);
 		return -EINVAL;
 	}
--- a/nss_pm.c
+++ b/nss_pm.c
@@ -51,20 +51,6 @@ static struct msm_bus_paths nss_gmac_bw_
 				/* 266 MHz to DDR, 20 MHz to TCM  */
 };
 
-/*
- * Bus vector table for Crypto driver
- */
-static struct msm_bus_paths nss_crypto_bw_level_tbl[NSS_PM_PERF_MAX_LEVELS] = {
-		[NSS_PM_PERF_LEVEL_SUSPEND] =  CRYPTO_BW_MBPS(0, 0),
-				/* 0 MHz to DDR, 0 MHz to TCM */
-		[NSS_PM_PERF_LEVEL_IDLE] =  CRYPTO_BW_MBPS(133, 5),
-				/* 133 MHz to DDR, 5 MHz to TCM */
-		[NSS_PM_PERF_LEVEL_NOMINAL] =  CRYPTO_BW_MBPS(200, 400),
-				/* 200 MHz to DDR, 10 MHz to TCM */
-		[NSS_PM_PERF_LEVEL_TURBO] = CRYPTO_BW_MBPS(266, 533),
-				/* 266 MHz to DDR, 20 MHz to TCM  */
-};
-
 #ifdef  NSS_PM_NETAP_GMAC_SCALING
 
 /*
@@ -101,7 +87,7 @@ static struct msm_bus_paths nss_netap_bw
 #endif
 
 /*
- * Bus Driver Platform data for GMAC, Crypto and Netap clients
+ * Bus Driver Platform data for GMAC, and Netap clients
  */
 static struct msm_bus_scale_pdata nss_bus_scale[] = {
 	[NSS_PM_CLIENT_GMAC] = {
@@ -111,13 +97,6 @@ static struct msm_bus_scale_pdata nss_bu
 		.name = "qca-nss-gmac",
 	},
 
-	[NSS_PM_CLIENT_CRYPTO] = {
-		.usecase = nss_crypto_bw_level_tbl,
-		.num_usecases = ARRAY_SIZE(nss_crypto_bw_level_tbl),
-		.active_only = 1,
-		.name = "qca-nss-crypto",
-	},
-
 	[NSS_PM_CLIENT_NETAP] = {
 		.usecase = nss_netap_bw_level_tbl,
 		.num_usecases = ARRAY_SIZE(nss_netap_bw_level_tbl),
@@ -239,10 +218,6 @@ void *nss_pm_client_register(nss_pm_clie
 			pm_dentry = debugfs_create_dir("gmac" , ctx.pm_dentry);
 			break;
 
-		case NSS_PM_CLIENT_CRYPTO:
-			pm_dentry = debugfs_create_dir("crypto" , ctx.pm_dentry);
-			break;
-
 		case NSS_PM_CLIENT_NETAP:
 			pm_dentry = debugfs_create_dir("netap" , ctx.pm_dentry);
 			break;
@@ -383,24 +358,6 @@ nss_pm_interface_status_t nss_pm_set_per
 		}
 	}
 
-	if (pm_client->client_id == NSS_PM_CLIENT_CRYPTO) {
-		if ((lvl == NSS_PM_PERF_LEVEL_TURBO) && (ctx.turbo_support == true))  {
-			/*
-			 * For Turbo mode, set Crypto core and
-			 * Fabric  port clocks to 213 MHz
-			 */
-			writel(0x23, CE5_ACLK_SRC0_NS);
-			writel(0x23, CE5_HCLK_SRC0_NS);
-			writel(0x23, CE5_CORE_CLK_SRC0_NS);
-
-			writel(0x2, CE5_ACLK_SRC_CTL);
-			writel(0x2, CE5_HCLK_SRC_CTL);
-			writel(0x2, CE5_CORE_CLK_SRC_CTL);
-		} else {
-			lvl = NSS_PM_PERF_LEVEL_NOMINAL;
-		}
-	}
-
 	/* Update bandwidth if request has changed. This may sleep. */
 	ret = msm_bus_scale_client_update_request(pm_client->bus_perf_client, lvl);
 	if (ret) {
